modules/em/src/FitRestraintBayesEM3D.cpp

```diff
@@ -0,0 +1,86 @@
+/**
+ *  \file FitRestraintBayesEM3D.cpp
+ *  \brief Calculate the Bayesian score and derivative
+ *   based on fit to an EM density map. This restraint differs from em::FitRestraint
+ *   because it does not use a cross_correlation_coefficient as score but a Bayesian
+ *   similarity measure.
+ *
+ *  Copyright 2007-2020 IMP Inventors. All rights reserved.
+ *
+ */
+#include <IMP/em/FitRestraintBayesEM3D.h>
+#include <IMP/em/MRCReaderWriter.h>
+#include <IMP/em/envelope_penetration.h>
+#include <IMP/core/internal/evaluate_distance_pair_score.h>
+#include <IMP/core/HarmonicUpperBound.h>
+#include <IMP/atom/pdb.h>
+#include "IMP/container_macros.h"
+#include <IMP/log.h>
+
+IMPEM_BEGIN_NAMESPACE
+
+FitRestraintBayesEM3D::FitRestraintBayesEM3D(
+  ParticlesTemp ps, 
+  DensityMap *em_map, 
+  FloatKey weight_key,
+  bool use_rigid_bodies, 
+  double sigma) 
+: Restraint(IMP::internal::get_model(ps), "Fit restraint BayesEM3D %1%")
+{
+
+  target_dens_map_ = em_map;
+  ps_=get_as<Particles>(ps);
+  //add_particles(ps);
+
+  resolution_ = target_dens_map_->get_header()->get_resolution();
+  voxel_size_ = target_dens_map_->get_spacing();
+
+  weight_key_ = weight_key;
+  sigma_ = sigma;
+
+  score_ = 0.;
+  dv_.insert(dv_.end(), ps_.size(), algebra::Vector3D(0., 0., 0.));
+
+}
+
+double FitRestraintBayesEM3D::unprotected_evaluate(DerivativeAccumulator *accum) const {
+
+  bool calc_deriv = accum ? true : false;
+  double score;
+
+  std::pair<double, algebra::Vector3Ds> vals =
+  IMP::em::bayesem3d_calc_score_and_derivative(
+    const_cast<DensityMap *> (target_dens_map_.get()),
+    ps_,
+    resolution_,
+    sigma_);
+
+  const_cast<FitRestraintBayesEM3D *>(this)->score_ = vals.first;
+  const_cast<FitRestraintBayesEM3D *>(this)->dv_ = vals.second;
```

Comment:
Ugly! Declare `score_` and `dv_` in your class to be `mutable`; then you can modify them in a `const` method.

Reply:
@benmwebb said: "Normally you can't change things in the class,  because you are in a `const `method. Instead, just use `score_ = vals.first;` as per normal. In the  corresponding header, declare the member as `mutable double score_`. The `mutable` attribute means you can change it even in a `const` method. It is the usual design pattern for members used for caches and the like. See [mutable](https://riptutorial.com/cplusplus/example/18647/mutable) for an example."

Fixed as suggested.